mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -0,0 +1,310 @@
+                    builder.create<scf::YieldOp>(loc, adjointOp.getResult());
+                })
+            .getResult(0);
+    std::vector<Value> argsAndRegMeasurement = {fnFoldedOp.getArguments().begin(),
+                                                fnFoldedOp.getArguments().end()};
+    argsAndRegMeasurement.pop_back();
+    argsAndRegMeasurement.push_back(loopedQreg);
+    ValueRange funcFolded =
+        rewriter.create<func::CallOp>(loc, fnWithMeasurementsOp, argsAndRegMeasurement)
+            .getResults();
+    rewriter.create<func::ReturnOp>(loc, funcFolded);
+    return SymbolRefAttr::get(ctx, fnFoldedName);
+}
+FlatSymbolRefAttr ZneLowering::getOrInsertQuantumAlloc(Location loc, PatternRewriter &rewriter,
+                                                       mitigation::ZneOp op)
+{
+    // Quantum Alloc function
+    MLIRContext *ctx = rewriter.getContext();
+    OpBuilder::InsertionGuard guard(rewriter);
+    ModuleOp moduleOp = op->getParentOfType<ModuleOp>();
+    Type qregType = quantum::QuregType::get(rewriter.getContext());
+
+    std::string fnAllocName = op.getCallee().str() + ".quantumAlloc";
+    if (moduleOp.lookupSymbol<func::FuncOp>(fnAllocName)) {
+        return SymbolRefAttr::get(ctx, fnAllocName);
+    }
+    Type i64Type = rewriter.getI64Type();
+    FunctionType fnAllocType = FunctionType::get(ctx, /*inputs=*/
+                                                 i64Type,
+                                                 /*outputs=*/qregType);
+    func::FuncOp fnAlloc = rewriter.create<func::FuncOp>(loc, fnAllocName, fnAllocType);
+    fnAlloc.setPrivate();
+    Block *allocBloc = fnAlloc.addEntryBlock();
+    rewriter.setInsertionPointToStart(allocBloc);
+    Value nQubits = allocBloc->getArgument(0);
+    IntegerAttr intAttr;
+    auto qreg = rewriter.create<quantum::AllocOp>(loc, qregType, nQubits, intAttr);
+    rewriter.create<func::ReturnOp>(loc, qreg.getResult());
+    return SymbolRefAttr::get(ctx, fnAllocName);
+}
+FlatSymbolRefAttr ZneLowering::getOrInsertFnWithoutMeasurements(Location loc,
+                                                                PatternRewriter &rewriter,
+                                                                mitigation::ZneOp op)
+{
+    MLIRContext *ctx = rewriter.getContext();
+    OpBuilder::InsertionGuard guard(rewriter);
+    ModuleOp moduleOp = op->getParentOfType<ModuleOp>();
+    Type qregType = quantum::QuregType::get(rewriter.getContext());
+    func::FuncOp fnOp = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
+    TypeRange originalTypes = op.getArgs().getTypes();
+
+    std::string fnWithoutMeasurementsName = op.getCallee().str() + ".withoutMeasurements";
+    if (moduleOp.lookupSymbol<func::FuncOp>(fnWithoutMeasurementsName)) {
+        return SymbolRefAttr::get(ctx, fnWithoutMeasurementsName);
+    }
+
+    SmallVector<Type> typesWithoutMeasurements = {originalTypes.begin(), originalTypes.end()};
```

Comment:
I couldn't understand why you create `typesWithMeasurements` with the first and last args types. Could you please elaborate?

Reply:
That's a mistake, I corrected it thanks!